### DISCUSSION TITLE
crank-nicolson fails to converge at the initial step

### DISCUSSION THREAD

**Discussion by ananthanarasimhanj** [idaholab/moose#27547] [Q&A General]
if I use crank nicolson, it fails to converge in the first timestep. What can be evaluated in-general to address this issue? without specifying any scheme, the simulation runs fine.  I can share the input file here if that helps (but input file is for CRANE - a MOOSE based).

**Comment by GiudGiud** [comment]
Hello

Which equation are you trying to solve?
Is CRANE open-source?
You can try the troubleshooting steps here:
https://mooseframework.inl.gov/moose/application_usage/failed_solves.html

Guillaume

**Reply by ananthanarasimhanj** [reply]
Trying to solve "ChemicalReactions" network. Yes, CRANE is open source built on MOOSE framework.

If I disable "crank-nicolson", the solver runs fine.

**Reply by GiudGiud** [reply]
What about bdf2?

Do all the nonlinear equations involved have a time derivative term? That s a known issue for our implementation of crank Nicholson

**Reply by ananthanarasimhanj** [reply]
Yes, all the equations have a time derivative term. 

I just tried with bdf2, and it works fine and runs very fast also. 

But the physical end time at which I get steady state convergence changes, and the final solution (density of each species) varies slightly with or without bdf2. Is that fine? what is the default time-integrator if I don't specify any? is there a recommended time-integrator for chemical reaction network?

**Reply by GiudGiud** [reply]
The default time integrator is implicit (first order) Euler.
Some differences in the results is expected if you have not converged the time step size.
I don't know about recommendations for time integration of chemical reaction network. The selection of a numerical scheme is typically done based on what is done elsewhere in the litterature of that field.

**Reply by ananthanarasimhanj** [reply]
Thanks! I will look into the literature. 

I guess the fast run I observed now is due to disabling the growth factor in the TimeStepper. Here is the executioner block I use now. 


```
[Executioner]
  type = Transient
  end_time = 1e3#s 
  #solve_type = 'newton'
  solve_type = 'pjfnk'
  #scheme = crank-nicolson
  scheme = bdf2
  petsc_options_iname = '-snes_linesearch_type'
  petsc_options_value = 'basic'  
  automatic_scaling = true
  steady_state_detection = true
  compute_scaling_once = false
  #nl_rel_tol = 1e-5
  [TimeStepper]
    type = IterationAdaptiveDT
    dt = 1e-3   
    growth_factor = 1.01    
  []

[]
```

**Reply by GiudGiud** [reply]
The growth factor helps the time stepper increase the time step. This can be useful when simply searching for steady state conditions.
Of course for a transient one does not want to grow the time step too much, as it results in a loss of accuracy

**Reply by ananthanarasimhanj** [reply]
Yes, if I disable growth_factor, the simulation is faster and achieve steady state condition quick, but with lose of accuracy. So, I have set it to 1.01. I think the default is 2. 

So, I guess I can proceed with bdf2 instead of CN?

**Reply by lindsayad** [reply]
You might want to look at [this CFD thread](https://www.cfd-online.com/Forums/main/250505-second-order-backward-differencing-vs-crank-nicolson.html#post852150) regarding BDF2 vs. Crank-Nicolson. I will quote one comment here:

> 1) It [BDF2] is, to the best of my knowledge, the only second order implicit A-stable scheme

**Reply by lindsayad** [reply]
another quote from that thread
> As a practitioner, the very point is that BDF2 is more stable than CN and discretizes the time derivative at n+1.

**Comment by ananthanarasimhanj** [comment]
I was trying to use lstabledirk4 ([MOOSE object link](https://mooseframework.inl.gov/syntax/Executioner/TimeIntegrator/index.html)), but I get the following error message. And what does the dirk in the below error message refer to? 4th order? 


`Invalid option "lstabledirk4" in MooseEnum.  Valid options (not case-sensitive) are "implicit-euler explicit-euler crank-nicolson bdf2 explicit-midpoint dirk explicit-tvd-rk-2 newmark-beta".`

**Reply by GiudGiud** [reply]
the shorthand syntax does not support it
add a `[TimeIntegrator]` sub-block to the Executioner blokc

**Reply by ananthanarasimhanj** [reply]
Thanks! It fixed. 

Also, whether MOOSE allows high-order (order > 5th) adaptive dt integration?

**Reply by GiudGiud** [reply]
5th order has not been implemented but I dont see why we could not implement it if needed
